hazelcast/src/main/java/com/hazelcast/query/impl/GlobalIndexPartitionTracker.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.query.impl;
+
+import com.hazelcast.internal.util.collection.PartitionIdSet;
+
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Tracker of indexed partition for global indexes.
+ * <p>
+ * Used by {@link InternalIndex} to keep track of indexed partitions and decide whether the query
+ * using this index could be executed or not.
+ */
+public class GlobalIndexPartitionTracker {
+
+    private static final long VERSION_INITIAL = 0;
+
+    /** Lock to serialize updates to the state. */
+    private final ReentrantLock lock = new ReentrantLock();
+
+    /** Number of partitions in the cluster. */
+    private final int partitionCount;
+
+    /** Current state. */
+    private final AtomicReference<State> state;
+
+    public GlobalIndexPartitionTracker(int partitionCount) {
+        this.partitionCount = partitionCount;
+
+        state = new AtomicReference<>(new State(VERSION_INITIAL, new PartitionIdSet(partitionCount), 0));
+    }
+
+    public Long getPartitionStamp(PartitionIdSet expectedPartitionIds) {
+        State state0 = state.get();
+
+        if (state0.pending > 0 || !state0.indexedPartitions.equals(expectedPartitionIds)) {
+            return null;
+        }
+
+        return state0.version;
+    }
+
+    public boolean validatePartitionStamp(long version) {
+        return state.get().version == version;
+    }
+
+    public boolean isMarked(int partitionId) {
+        return state.get().indexedPartitions.contains(partitionId);
```

Comment:
Since we don't check the `pending` count here, is it possible we say the index partition is marked/indexed while it's being (un)populated?

Reply:
Yes, this is a kind of counter-intuitive for the case when we begin the de-indexing of the indexed partition. This is not a problem for SQL, because it relies on stamps. But could be a problem for the old predicate API - there is definitively a window of inconsistency there. 

I will create a follow-up issue to fix the predicate API as well. This is not urgent, because it is not a regression, but a bug that sits there for years.